File: apps/api/repositories/bookings/bookings.go
package bookings

import (
	"context"
	"errors"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/kinsittr/kinsittr-api/models"
)

var (
	ErrBookingAlreadyExists       = errors.New("booking_already_exists")
	ErrNannyTimeUnavailable       = errors.New("nanny_time_unavailable")
	ErrPendingChangeRequestExists = errors.New("pending_change_request_exists")
)

type BookingRecord struct {
	models.Booking
	ConversationID        string `json:"conversation_id,omitempty"`
	NannyDisplayName      string `json:"nanny_display_name"`
	NannyCity             string `json:"nanny_city"`
	NannyProvince         string `json:"nanny_province"`
	ParentDisplayName     string `json:"parent_display_name"`
	ParentCity            string `json:"parent_city"`
	ParentProvince        string `json:"parent_province"`
	PaymentStatus         string `json:"payment_status,omitempty"`
	PaymentFailureMessage string `json:"payment_failure_message,omitempty"`
	StripePaymentIntentID string `json:"stripe_payment_intent_id,omitempty"`
	StripeChargeID        string `json:"stripe_charge_id,omitempty"`
	StripeRefundID        string `json:"stripe_refund_id,omitempty"`
}

type ListBookingsFilter struct {
	Page     int
	Limit    int
	Status   models.BookingStatus
	DateFrom *time.Time
	DateTo   *time.Time
}

type BookingsRepository interface {
	HasParentActiveBookingWithNanny(ctx context.Context, parentProfileID, nannyProfileID uuid.UUID, startTime time.Time, duration int) (bool, error)
	HasNannyApprovedBookingConflict(ctx context.Context, nannyProfileID uuid.UUID, startTime time.Time, duration int) (bool, error)
	HasNannyApprovedBookingConflictExcluding(ctx context.Context, nannyProfileID uuid.UUID, startTime time.Time, duration int, excludeBookingID uuid.UUID) (bool, error)
	CreateBooking(ctx context.Context, booking models.Booking) (models.Booking, error)
	ListParentBookings(ctx context.Context, parentProfileID uuid.UUID, filter ListBookingsFilter) ([]BookingRecord, int, error)
	GetParentBookingByID(ctx context.Context, parentProfileID, bookingID uuid.UUID) (BookingRecord, error)
	CancelParentBooking(ctx context.Context, parentProfileID, bookingID uuid.UUID) (BookingRecord, error)
	ListNannyBookings(ctx context.Context, nannyProfileID uuid.UUID, filter ListBookingsFilter) ([]BookingRecord, int, error)
	GetNannyBookingByID(ctx context.Context, nannyProfileID, bookingID uuid.UUID) (BookingRecord, error)
	ApproveNannyBooking(ctx context.Context, nannyProfileID, bookingID uuid.UUID) (BookingRecord, error)
	ApproveNannyBookingWithConversation(ctx context.Context, nannyProfileID, bookingID uuid.UUID) (BookingRecord, error)
	DeclineNannyBooking(ctx context.Context, nannyProfileID, bookingID uuid.UUID) (BookingRecord, error)
	CreateBookingChangeRequest(ctx context.Context, request models.BookingChangeRequest) (models.BookingChangeRequest, error)
	ListBookingChangeRequests(ctx context.Context, bookingID uuid.UUID) ([]models.BookingChangeRequest, error)
	GetBookingChangeRequestByID(ctx context.Context, bookingID, requestID uuid.UUID) (models.BookingChangeRequest, error)
	AcceptBookingChangeRequest(ctx context.Context, bookingID, requestID uuid.UUID, responseNote string) (BookingRecord, models.BookingChangeRequest, error)
	DeclineBookingChangeRequest(ctx context.Context, bookingID, requestID uuid.UUID, responseNote string) (models.BookingChangeRequest, error)
	CompleteNannyBooking(ctx context.Context, nannyProfileID, bookingID uuid.UUID) (BookingRecord, error)
}

var BookingsRepo BookingsRepository

func InitBookingsRepo(db *pgxpool.Pool) {
	BookingsRepo = newPgRepository(db)
}

File: apps/web/src/types/api/bookings.ts
export interface CreateBookingPayload {
  nanny_id: string;
  date: string;
  start_time: string;
  timezone_offset_minutes: number;
  duration: number;
}

export type BookingStatus = "pending" | "approved" | "declined" | "cancelled" | "completed";
export type PaymentStatus =
  | "requires_payment_method"
  | "requires_confirmation"
  | "requires_action"
  | "processing"
  | "succeeded"
  | "failed"
  | "cancelled"
  | "refunded";

export interface Booking {
  id: string;
  conversation_id?: string;
  parent_profile_id: string;
  nanny_profile_id: string;
  parent_display_name?: string;
  parent_city?: string;
  parent_province?: string;
  parent_num_children?: number;
  nanny_display_name?: string;
  nanny_city?: string;
  nanny_province?: string;
  date: string;
  start_time: string;
  duration: number;
  total_amount: number;
  status: BookingStatus;
  payment_status?: PaymentStatus | "";
  payment_failure_message?: string;
  stripe_payment_intent_id?: string;
  stripe_charge_id?: string;
  stripe_refund_id?: string;
  created_at: string;
  updated_at: string;
}

export interface BookingListData {
  items: Booking[];
  page: number;
  limit: number;
  total: number;
}

export interface ListBookingsParams {
  page?: number;
  limit?: number;
  status?: BookingStatus;
  date_from?: string;
  date_to?: string;
}
